File: js/task3c/S3_PIPE_06.js
function pipeSafe(...fns) {
  return function(initialValue) {
    let value = initialValue;
    for (const fn of fns) {
      try {
        value = fn(value);
      } catch (error) {
        return { ok: false, error };
      }
    }
    return { ok: true, value };
  };
}

const safePipeline = pipeSafe(
  x => x + 1,
  x => {
    if (x > 2) throw new Error("Too big!");
    return x * 2;
  },
  x => x - 1
);
console.log(safePipeline(0)); 
console.log(safePipeline(2)); 
console.log(safePipeline(3));

File: python/task3c/S3_PIPE_06.py
def pipe_safe(*fns):
  def inner(initial_value):
      value = initial_value
      for fn in fns:
          try:
              value = fn(value)
          except Exception as error:
              return {"ok": False, "error": error}
      return {"ok": True, "value": value}
  return inner


safe_pipeline = pipe_safe(
  lambda x: x + 1,
  lambda x: (_ for _ in ()).throw(Exception("Too big!")) if x > 2 else x * 2,
  lambda x: x - 1
)

print(safe_pipeline(0))
print(safe_pipeline(2))
print(safe_pipeline(3))
